### PR TITLE
Flinging continuously should have one scroll start/end pair

### LINKED ISSUE #2430: Scroll Events are still wonky:
Example logging of pointer and scroll start/ends:

```
$ flutter start && adb logcat -s -vtime flutter:D
Using pre-built SkyShell.apk.
Starting lib/main.dart on Pixel C...
--------- beginning of main
--------- beginning of system
03-04 13:48:38.873 I/flutter (23259): onPointerDown
03-04 13:48:38.928 I/flutter (23259): onPointerUp
03-04 13:48:38.939 I/flutter (23259): TRIGGERING SCROLL
03-04 13:48:38.959 I/flutter (23259): onScrollStart
03-04 13:48:39.365 I/flutter (23259): onScrollEnd
03-04 13:48:40.080 I/flutter (23259): onPointerDown
03-04 13:48:40.082 I/flutter (23259): onScrollStart
03-04 13:48:40.191 I/flutter (23259): onPointerUp
03-04 13:48:40.543 I/flutter (23259): onPointerDown
03-04 13:48:40.544 I/flutter (23259): onScrollEnd
03-04 13:48:40.632 I/flutter (23259): onScrollStart
03-04 13:48:40.681 I/flutter (23259): onPointerUp
03-04 13:48:41.077 I/flutter (23259): onPointerDown
03-04 13:48:41.086 I/flutter (23259): onScrollEnd
03-04 13:48:41.100 I/flutter (23259): onScrollStart
03-04 13:48:41.148 I/flutter (23259): onPointerUp
03-04 13:48:41.581 I/flutter (23259): onPointerDown
03-04 13:48:41.582 I/flutter (23259): onScrollEnd
03-04 13:48:41.648 I/flutter (23259): onScrollStart
03-04 13:48:41.679 I/flutter (23259): onPointerUp
03-04 13:48:42.233 I/flutter (23259): onPointerDown
03-04 13:48:42.234 I/flutter (23259): onScrollEnd
03-04 13:48:42.284 I/flutter (23259): onScrollStart
03-04 13:48:42.324 I/flutter (23259): onPointerUp
03-04 13:48:42.865 I/flutter (23259): onPointerDown
03-04 13:48:42.873 I/flutter (23259): onScrollEnd
03-04 13:48:42.886 I/flutter (23259): onScrollStart
03-04 13:48:42.953 I/flutter (23259): onPointerUp
03-04 13:48:43.556 I/flutter (23259): onPointerDown
03-04 13:48:43.556 I/flutter (23259): onScrollEnd
03-04 13:48:43.666 I/flutter (23259): onScrollStart
03-04 13:48:43.693 I/flutter (23259): onPointerUp
03-04 13:48:44.914 I/flutter (23259): onPointerDown
03-04 13:48:44.970 I/flutter (23259): onScrollEnd
03-04 13:48:44.972 I/flutter (23259): onScrollStart
03-04 13:48:44.972 I/flutter (23259): onScrollEnd
03-04 13:48:45.002 I/flutter (23259): onScrollStart
03-04 13:48:45.002 I/flutter (23259): onScrollEnd
03-04 13:48:45.003 I/flutter (23259): onScrollStart
03-04 13:48:45.003 I/flutter (23259): onScrollEnd
03-04 13:48:45.003 I/flutter (23259): onPointerUp
03-04 13:48:45.004 I/flutter (23259): onScrollStart
03-04 13:48:45.239 I/flutter (23259): onPointerDown
03-04 13:48:45.241 I/flutter (23259): onScrollEnd
03-04 13:48:45.243 I/flutter (23259): onScrollStart
03-04 13:48:45.243 I/flutter (23259): onScrollEnd
03-04 13:48:45.272 I/flutter (23259): onScrollStart
03-04 13:48:45.272 I/flutter (23259): onScrollEnd
03-04 13:48:45.275 I/flutter (23259): onScrollStart
03-04 13:48:45.275 I/flutter (23259): onScrollEnd
03-04 13:48:45.276 I/flutter (23259): onPointerUp
03-04 13:48:45.277 I/flutter (23259): onScrollStart
03-04 13:48:46.031 I/flutter (23259): onPointerDown
03-04 13:48:46.088 I/flutter (23259): onScrollEnd
03-04 13:48:46.091 I/flutter (23259): onScrollStart
03-04 13:48:46.091 I/flutter (23259): onScrollEnd
03-04 13:48:46.100 I/flutter (23259): onScrollStart
03-04 13:48:46.100 I/flutter (23259): onScrollEnd
03-04 13:48:46.119 I/flutter (23259): onPointerUp
03-04 13:48:46.120 I/flutter (23259): onScrollStart
03-04 13:48:46.493 I/flutter (23259): onPointerDown
03-04 13:48:46.495 I/flutter (23259): onScrollEnd
03-04 13:48:46.498 I/flutter (23259): onScrollStart
03-04 13:48:46.525 I/flutter (23259): onPointerUp
03-04 13:48:47.661 I/flutter (23259): onPointerDown
03-04 13:48:47.717 I/flutter (23259): onScrollEnd
03-04 13:48:47.719 I/flutter (23259): onScrollStart
03-04 13:48:47.719 I/flutter (23259): onScrollEnd
03-04 13:48:47.727 I/flutter (23259): onPointerUp
03-04 13:48:47.728 I/flutter (23259): onScrollStart
03-04 13:48:48.322 I/flutter (23259): onPointerDown
03-04 13:48:48.323 I/flutter (23259): onScrollEnd
03-04 13:48:48.380 I/flutter (23259): onScrollStart
03-04 13:48:48.407 I/flutter (23259): onPointerUp
03-04 13:48:48.832 I/flutter (23259): onPointerDown
03-04 13:48:48.832 I/flutter (23259): onScrollEnd
03-04 13:48:48.856 I/flutter (23259): onScrollStart
03-04 13:48:48.897 I/flutter (23259): onPointerUp
03-04 13:48:49.324 I/flutter (23259): onPointerDown
03-04 13:48:49.379 I/flutter (23259): onScrollEnd
03-04 13:48:49.381 I/flutter (23259): onScrollStart
03-04 13:48:49.381 I/flutter (23259): onScrollEnd
03-04 13:48:49.407 I/flutter (23259): onScrollStart
03-04 13:48:49.407 I/flutter (23259): onScrollEnd
03-04 13:48:49.407 I/flutter (23259): onPointerUp
03-04 13:48:49.408 I/flutter (23259): onScrollStart
03-04 13:48:49.840 I/flutter (23259): onPointerDown
03-04 13:48:49.841 I/flutter (23259): onScrollEnd
03-04 13:48:49.881 I/flutter (23259): onScrollStart
03-04 13:48:49.919 I/flutter (23259): onPointerUp
03-04 13:48:50.269 I/flutter (23259): onPointerDown
03-04 13:48:50.269 I/flutter (23259): onScrollEnd
03-04 13:48:50.306 I/flutter (23259): onScrollStart
03-04 13:48:50.342 I/flutter (23259): onPointerUp
03-04 13:48:50.798 I/flutter (23259): onPointerDown
03-04 13:48:50.798 I/flutter (23259): onScrollEnd
03-04 13:48:50.806 I/flutter (23259): onScrollStart
03-04 13:48:50.807 I/flutter (23259): onScrollEnd
03-04 13:48:50.833 I/flutter (23259): onScrollStart
03-04 13:48:50.834 I/flutter (23259): onScrollEnd
03-04 13:48:50.834 I/flutter (23259): onScrollStart
03-04 13:48:50.835 I/flutter (23259): onScrollEnd
03-04 13:48:50.835 I/flutter (23259): onScrollStart
03-04 13:48:50.835 I/flutter (23259): onScrollEnd
03-04 13:48:50.841 I/flutter (23259): onScrollStart
03-04 13:48:50.841 I/flutter (23259): onScrollEnd
03-04 13:48:50.848 I/flutter (23259): onPointerUp
03-04 13:48:50.849 I/flutter (23259): onScrollStart
03-04 13:48:51.333 I/flutter (23259): onPointerDown
03-04 13:48:51.333 I/flutter (23259): onScrollEnd
03-04 13:48:51.357 I/flutter (23259): onScrollStart
03-04 13:48:51.392 I/flutter (23259): onPointerUp
03-04 13:48:51.807 I/flutter (23259): onPointerDown
03-04 13:48:51.808 I/flutter (23259): onScrollEnd
03-04 13:48:51.832 I/flutter (23259): onScrollStart
03-04 13:48:51.871 I/flutter (23259): onPointerUp
03-04 13:48:52.578 I/flutter (23259): onPointerDown
03-04 13:48:52.580 I/flutter (23259): onScrollEnd
03-04 13:48:52.587 I/flutter (23259): onScrollStart
03-04 13:48:52.587 I/flutter (23259): onScrollEnd
03-04 13:48:52.614 I/flutter (23259): onScrollStart
03-04 13:48:52.614 I/flutter (23259): onScrollEnd
03-04 13:48:52.615 I/flutter (23259): onPointerUp
03-04 13:48:52.615 I/flutter (23259): onScrollStart
03-04 13:48:52.647 I/flutter (23259): onScrollEnd
03-04 13:48:52.884 I/flutter (23259): onPointerDown
03-04 13:48:52.900 I/flutter (23259): onScrollStart
03-04 13:48:52.942 I/flutter (23259): onPointerUp
03-04 13:48:53.342 I/flutter (23259): onPointerDown
03-04 13:48:53.343 I/flutter (23259): onScrollEnd
03-04 13:48:53.350 I/flutter (23259): onScrollStart
03-04 13:48:53.401 I/flutter (23259): onPointerUp
03-04 13:48:53.577 I/flutter (23259): onPointerDown
03-04 13:48:53.577 I/flutter (23259): onScrollEnd
03-04 13:48:53.596 I/flutter (23259): onScrollStart
03-04 13:48:53.630 I/flutter (23259): onPointerUp
03-04 13:48:54.037 I/flutter (23259): onPointerDown
03-04 13:48:54.037 I/flutter (23259): onScrollEnd
03-04 13:48:54.054 I/flutter (23259): onScrollStart
03-04 13:48:54.087 I/flutter (23259): onPointerUp
03-04 13:48:54.412 I/flutter (23259): onPointerDown
03-04 13:48:54.418 I/flutter (23259): onScrollEnd
03-04 13:48:54.447 I/flutter (23259): onScrollStart
03-04 13:48:54.447 I/flutter (23259): onScrollEnd
03-04 13:48:54.448 I/flutter (23259): onScrollStart
03-04 13:48:54.448 I/flutter (23259): onScrollEnd
03-04 13:48:54.449 I/flutter (23259): onScrollStart
03-04 13:48:54.449 I/flutter (23259): onScrollEnd
03-04 13:48:54.452 I/flutter (23259): onScrollStart
03-04 13:48:54.452 I/flutter (23259): onScrollEnd
03-04 13:48:54.483 I/flutter (23259): onScrollStart
03-04 13:48:54.483 I/flutter (23259): onScrollEnd
03-04 13:48:54.484 I/flutter (23259): onScrollStart
03-04 13:48:54.484 I/flutter (23259): onScrollEnd
03-04 13:48:54.485 I/flutter (23259): onScrollStart
03-04 13:48:54.485 I/flutter (23259): onScrollEnd
03-04 13:48:54.486 I/flutter (23259): onPointerUp
03-04 13:48:54.487 I/flutter (23259): onScrollStart
03-04 13:48:54.916 I/flutter (23259): onPointerDown
03-04 13:48:54.916 I/flutter (23259): onScrollEnd
03-04 13:48:54.927 I/flutter (23259): onScrollStart
03-04 13:48:54.970 I/flutter (23259): onPointerUp
03-04 13:48:55.953 I/flutter (23259): onPointerDown
03-04 13:48:55.953 I/flutter (23259): onScrollEnd
03-04 13:48:55.961 I/flutter (23259): onPointerUp
03-04 13:48:55.962 I/flutter (23259): onScrollStart
03-04 13:48:55.969 I/flutter (23259): onScrollEnd
03-04 13:49:04.409 I/flutter (23259): onPointerDown
03-04 13:49:04.635 I/flutter (23259): onScrollStart
03-04 13:49:06.594 I/flutter (23259): onPointerUp
03-04 13:49:07.050 I/flutter (23259): onScrollEnd
```

### PR DESCRIPTION
**Pull request by abarth**
Previously we got confused and started sending start/end pairs for each tick of
the fling animation.

Fixes #2430

### CODE REVIEW

**Comment by abarth** [comment]
@apwilson @Hixie

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 1]
```diff
@@ -470,11 +463,12 @@ abstract class ScrollableState<T extends Scrollable> extends State<T> {
...
 
   void _handleDragStart(_) {
-    _dispatchOnScrollStartIfNeeded();
+    _startScroll();
   }
 
-  void _dispatchOnScrollStartIfNeeded() {
-    if (!_isBetweenOnScrollStartAndOnScrollEnd)
+  void _startScroll() {
+    ++_numberOfInProgressScrolls;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scrollable.dart]: += 1

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 2]
```diff
@@ -495,11 +488,12 @@ abstract class ScrollableState<T extends Scrollable> extends State<T> {
...
     double scrollVelocity = pixelDeltaToScrollOffset(velocity.pixelsPerSecond) / Duration.MILLISECONDS_PER_SECOND;
     // The gesture velocity properties are pixels/second, config min,max limits are pixels/ms
-    return fling(scrollVelocity.clamp(-kMaxFlingVelocity, kMaxFlingVelocity)).then(_dispatchOnScrollEndIfNeeded);
+    return fling(scrollVelocity.clamp(-kMaxFlingVelocity, kMaxFlingVelocity)).then(_endScroll);
   }
 
-  void _dispatchOnScrollEndIfNeeded(_) {
-    if (_isBetweenOnScrollStartAndOnScrollEnd)
+  void _endScroll([_]) {
+    --_numberOfInProgressScrolls;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scrollable.dart]: `-= 1`, to avoid the `--_` ASCII art...

**Comment by Hixie** [comment]
This is an interesting approach, but what does it mean for there to be multiple animations at once? Do they fight? Surely whenever you, e.g., start dragging, any active animations get canceled?

LGTM anyway.

**Comment by abarth** [comment]
> what does it mean for there to be multiple animations at once? Do they fight? Surely whenever you, e.g., start dragging, any active animations get canceled?

The situations where there are multiple animations in flight don't last very long.  When we start an animation, we stop any previous animation.  However, the code that would deliver the onScrollEnd callback is watching for animations to end using a Future.  That means we don't deliver the onScrollEnd until the end of the microtask, which means the animations "overlap" during this period.  By the time the microtask is over and we check for onScrollEnd, we will have started the new animation (skipping the onScrollStart) and therefore we won't deliver the onScrollEnd until that animation is done.

I'll clean up the above and put it as a comment for this field.

**Comment by Hixie** [comment]
+1 for that as a comment, that really clears it up, thanks.

**Pull request merged**